pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -137,27 +142,64 @@ func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
 	return labelNames, warnings, nil
 }
 
-type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
+type stringSliceFunc func(context.Context, storage.Querier) ([]string, storage.Warnings, error)
+
+type stringSliceFuncJob struct {
+	querier  storage.Querier
+	tenantID string
+	result   []string
+	warnings storage.Warnings
+}
 
 // mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
-// on a querier. It removes duplicates and sorts the result. It doesn't require
-// the output of the stringSliceFunc to be sorted, as results of LabelValues
-// are not sorted.
-//
-// TODO: Consider running stringSliceFunc calls concurrently
+// on per querier in parallel. It removes duplicates and sorts the result. It
+// doesn't require the output of the stringSliceFunc to be sorted, as results
+// of LabelValues are not sorted.
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+	var jobs = make([]interface{}, len(m.tenantIDs))
+
+	for pos := range m.tenantIDs {
+		jobs[pos] = &stringSliceFuncJob{
+			querier:  m.queriers[pos],
+			tenantID: m.tenantIDs[pos],
+		}
+	}
+
+	run := func(ctx context.Context, jobIntf interface{}) error {
+		job, ok := jobIntf.(*stringSliceFuncJob)
+		if !ok {
+			return fmt.Errorf("unexpected type %T", jobIntf)
+		}
+
+		var err error
+		job.result, job.warnings, err = f(ctx, job.querier)
+		if err != nil {
+			return errors.Wrapf(err, "error querying %s %s", rewriteLabelName(defaultTenantLabel), job.tenantID)
```

Comment:
**question:** Why rewrite the default tenant label here and below? If I'm not mistaken the user will be using the default tenant label with the underscores in their queries. Wouldn't it be better to return it formatted using the label name they use in their queries?

```suggestion
			return errors.Wrapf(err, "error querying %s %s", defaultTenantLabel, job.tenantID)
```

Reply:
I had mainly done this for clarity as the user is not necessarily using this label it in the query (@pstibrany suggested to not let it appear like a label selector) 

Currently the only value is `__tenant_id__` which will be rewritten into `tenant_id`. The PR #4117 will allow external code to set that label.

Reply:
Caught up with @jtlisi out-of-band and he was ok with keeping it like that